File: toolbox/src/main/java/io/prizy/toolbox/templating/MustacheTemplateCompiler.java
package io.prizy.toolbox.templating;

import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;

import com.github.mustachejava.DefaultMustacheFactory;
import com.github.mustachejava.MustacheFactory;
import lombok.experimental.UtilityClass;

/**
 * @author Nidhal Dogga
 * @created 4/25/2022 10:26 PM
 */


@UtilityClass
public class MustacheTemplateCompiler {

  private final MustacheFactory mustacheFactory = new DefaultMustacheFactory();

  public String compileTemplate(String template, Object data) throws IOException {
    var sw = new StringWriter();
    mustacheFactory
      .compile(String.format("templates/%s", template))
      .execute(sw, data)
      .flush();
    return sw.toString();
  }

  public String compileRaw(String rawContent, Object data) throws IOException {
    var sw = new StringWriter();
    mustacheFactory
      .compile(new StringReader(rawContent), "cpu.template.email")
      .execute(sw, data)
      .flush();
    return sw.toString();
  }

}

File: toolbox/src/main/kotlin/io/prizy/toolbox/templating/MustacheTemplateCompiler.kt
package io.prizy.toolbox.templating

import com.github.mustachejava.DefaultMustacheFactory
import com.github.mustachejava.MustacheFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.StringWriter

/**
 *  @author Nidhal Dogga
 *  @created 4/25/2022 10:26 PM
 */

object MustacheTemplateCompiler {

  private val mustacheFactory: MustacheFactory = DefaultMustacheFactory()

  suspend fun compile(template: String, data: Any): String = withContext(Dispatchers.IO) {
    StringWriter()
      .also { sw ->
        mustacheFactory
          .compile("templates/$template")
          .execute(sw, data)
          .flush()
      }
      .toString()
  }
}